Strict key checking issues in v1.16

It appears that https://github.com/taverntesting/tavern/pull/604 had a couple of unintended side effects:

1. The default strictness setting appears to have been changed to `off` instead of the expected `on` setting
2. Having the strict level at `list_any_order` no longer does strict key checking within nested objects

As a quick example, I have a dummy server that returns the following JSON:

```json
{
  "a": 1,
  "b": {
    "first": 10,
    "second": 20
  },
  "c": 2
}
```

Tavern file:

```yaml
---
test_name: strict test
stages:
  - name: Simple get
    request:
      url: http://localhost:8080
      method: GET
    response:
      status_code: 200
      json:
        a: 1
        b:
          first: 10
#          second: 20
        c: 2
```

This unexpectedly passes:

```
$ pytest -- test.tavern.yaml
============================= test session starts ==============================
platform linux -- Python 3.9.6, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /volume/temp
plugins: tavern-1.16.0
collected 1 item

test.tavern.yaml .                                                       [100%]

============================== 1 passed in 0.18s ===============================
```

Having `strict: -json:list_any_order` has the same result.

Adding `strict: - json:on` to the YAML causes it to fail, as I would've expected without having to specify it explicitly:

```
$ pytest -- test.tavern.yaml
============================= test session starts ==============================
platform linux -- Python 3.9.6, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /volume/temp
plugins: tavern-1.16.0
collected 1 item

test.tavern.yaml F                                                       [100%]

=================================== FAILURES ===================================
__________________ /volume/temp/test.tavern.yaml::strict test __________________
Format variables:

Source test stage (line 5):
  - name: Simple get
    request:
      url: http://localhost:8080
      method: GET
    response:
      status_code: 200
      json:
        a: 1
        b:
          first: 10

        c: 2

Formatted stage:
  name: Simple get
  request:
    method: GET
    url: http://localhost:8080
  response:
    json:
      a: 1
      b:
        first: 10
      c: 2
    status_code: 200

Errors:
E   tavern.util.exceptions.TestFailError: Test 'Simple get' failed:
    - Structure of returned data was different than expected  - Extra keys in response: {'second'} (expected["b"] = '{'first': 10}' (type = <class 'dict'>), actual["b"] = '{'first': 10, 'second': 20}' (type = <class 'dict'>))
------------------------------ Captured log call -------------------------------
ERROR    tavern.response.base:base.py:41 Structure of returned data was different than expected  - Extra keys in response: {'second'} (expected["b"] = '{'first': 10}' (type = <class 'dict'>), actual["b"] = '{'first': 10, 'second': 20}' (type = <class 'dict'>))
Traceback (most recent call last):
  File "/volume/temp/.venv/lib/python3.9/site-packages/tavern/util/dict_util.py", line 402, in check_keys_match_recursive
    assert actual_val == expected_val
AssertionError: assert {'a': 1, 'b':...: 20}, 'c': 2} == {'a': 1, 'b':...: 10}, 'c': 2}
  Omitting 2 identical items, use -vv to show
  Differing items:
  {'b': {'first': 10, 'second': 20}} != {'b': {'first': 10}}
  Use -v to get the full diff

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/volume/temp/.venv/lib/python3.9/site-packages/tavern/util/dict_util.py", line 402, in check_keys_match_recursive
    assert actual_val == expected_val
AssertionError: assert {'first': 10, 'second': 20} == {'first': 10}
  Omitting 1 identical items, use -vv to show
  Left contains 1 more item:
  {'second': 20}
  Use -v to get the full diff

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/volume/temp/.venv/lib/python3.9/site-packages/tavern/response/base.py", line 89, in recurse_check_key_match
    check_keys_match_recursive(expected_block, block, [], strict)
  File "/volume/temp/.venv/lib/python3.9/site-packages/tavern/util/dict_util.py", line 460, in check_keys_match_recursive
    check_keys_match_recursive(
  File "/volume/temp/.venv/lib/python3.9/site-packages/tavern/util/dict_util.py", line 445, in check_keys_match_recursive
    raise exceptions.KeyMismatchError(full_msg) from e
tavern.util.exceptions.KeyMismatchError: Structure of returned data was different than expected  - Extra keys in response: {'second'} (expected["b"] = '{'first': 10}' (type = <class 'dict'>), actual["b"] = '{'first': 10, 'second': 20}' (type = <class 'dict'>))
=========================== short test summary info ============================
FAILED test.tavern.yaml::strict test
============================== 1 failed in 0.21s ===============================
```

If I change the Tavern YAML to comment out the `c` key from the response:

```yaml
---
test_name: strict test
stages:
  - name: Simple get
    request:
      url: http://localhost:8080
      method: GET
    response:
      status_code: 200
      json:
        a: 1
        b:
          first: 10
#          second: 20
#        c: 2
```

This passes without any `strict` key and fails in both `list_any_order` and `on`.

--- a/tavern/util/dict_util.py
+++ b/tavern/util/dict_util.py
@@ -396,7 +396,7 @@ def check_keys_match_recursive(expected_val, actual_val, keys, strict=True):
             issubclass(actual_type, type(expected_val))
         )
 
-    strict, strict_setting = extract_strict_setting(strict)
+    strict_bool, strict_setting = extract_strict_setting(strict)
 
     try:
         assert actual_val == expected_val
@@ -441,12 +441,12 @@ def check_keys_match_recursive(expected_val, actual_val, keys, strict=True):
 
                 # If there are more keys in 'expected' compared to 'actual',
                 # this is still a hard error and we shouldn't continue
-                if extra_expected_keys or strict:
+                if extra_expected_keys or strict_bool:
                     raise exceptions.KeyMismatchError(full_msg) from e
                 else:
                     logger.debug(
                         "Mismatch in returned data, continuing due to strict=%s: %s",
-                        strict,
+                        strict_bool,
                         full_msg,
                         exc_info=True,
                     )
@@ -458,16 +458,16 @@ def check_keys_match_recursive(expected_val, actual_val, keys, strict=True):
             for key in to_recurse:
                 try:
                     check_keys_match_recursive(
-                        expected_val[key], actual_val[key], keys + [key], strict_setting
+                        expected_val[key], actual_val[key], keys + [key], strict
                     )
                 except KeyError:
                     logger.debug(
                         "Skipping comparing missing key '%s' due to strict=%s",
                         key,
-                        strict,
+                        strict_bool,
                     )
         elif isinstance(expected_val, list):
-            if not strict:
+            if not strict_bool:
                 missing = []
 
                 actual_iter = iter(actual_val)
@@ -492,7 +492,7 @@ def check_keys_match_recursive(expected_val, actual_val, keys, strict=True):
                         # Found one - check if it matches
                         try:
                             check_keys_match_recursive(
-                                e_val, current_response_val, keys + [i], strict_setting
+                                e_val, current_response_val, keys + [i], strict
                             )
                         except exceptions.KeyMismatchError:
                             # Doesn't match what we're looking for
@@ -522,9 +522,7 @@ def check_keys_match_recursive(expected_val, actual_val, keys, strict=True):
 
                 for i, (e_val, a_val) in enumerate(zip(expected_val, actual_val)):
                     try:
-                        check_keys_match_recursive(
-                            e_val, a_val, keys + [i], strict_setting
-                        )
+                        check_keys_match_recursive(e_val, a_val, keys + [i], strict)
                     except exceptions.KeyMismatchError as sub_e:
                         # This should _ALWAYS_ raise an error (unless the reason it didn't match was the
                         # 'anything' sentinel), but it will be more obvious where the error came from

--- a/tavern/util/strict_util.py
+++ b/tavern/util/strict_util.py
@@ -119,6 +119,9 @@ class StrictLevel:
 
 def extract_strict_setting(strict):
     """Takes either a bool, StrictOption, or a StrictSetting and return the bool representation and StrictSetting representation"""
+
+    logger.debug("Parsing a '%s': %s", type(strict), strict)
+
     if isinstance(strict, StrictSetting):
         strict_setting = strict
         strict = strict == StrictSetting.ON

--- a/tests/integration/server.py
+++ b/tests/integration/server.py
@@ -1,6 +1,7 @@
 import base64
 import itertools
 import json
+import math
 import mimetypes
 import os
 import time
@@ -8,7 +9,6 @@ from urllib.parse import unquote_plus
 import uuid
 
 from flask import Flask, Response, jsonify, redirect, request
-import math
 
 app = Flask(__name__)
 
@@ -315,3 +315,8 @@ def return_with_dot():
 @app.route("/uuid/v4", methods=["GET"])
 def get_uuid_v4():
     return jsonify({"uuid": uuid.uuid4()}), 200
+
+
+@app.route("/707-regression", methods=["GET"])
+def get_707():
+    return jsonify({"a": 1, "b": {"first": 10, "second": 20}, "c": 2})

--- a/tests/integration/test_strict_key_checks.tavern.yaml
+++ b/tests/integration/test_strict_key_checks.tavern.yaml
@@ -557,3 +557,23 @@ stages:
       strict: true
       json:
         - c
+
+---
+
+test_name: strict test
+
+_xfail: run
+
+stages:
+- name: Simple get
+  request:
+    url: "{global_host}/707-regression"
+    method: GET
+  response:
+    status_code: 200
+    json:
+      a: 1
+      b:
+        first: 10
+        # second: 20
+      c: 2

--- a/tests/unit/test_strict_util.py
+++ b/tests/unit/test_strict_util.py
@@ -9,6 +9,10 @@ from tavern.util.strict_util import StrictOption, StrictSetting, extract_strict_
 def test_extract_strict_setting_true(strict):
     as_bool, as_setting = extract_strict_setting(strict)
     assert as_bool is True
+    if isinstance(strict, StrictSetting):
+        assert as_setting == strict
+    if isinstance(strict, StrictOption):
+        assert as_setting == strict.setting
 
 
 @pytest.mark.parametrize(
@@ -25,3 +29,7 @@ def test_extract_strict_setting_true(strict):
 def test_extract_strict_setting_false(strict):
     as_bool, as_setting = extract_strict_setting(strict)
     assert as_bool is False
+    if isinstance(strict, StrictSetting):
+        assert as_setting == strict
+    if isinstance(strict, StrictOption):
+        assert as_setting == strict.setting

